src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -72,13 +85,163 @@ partition_allocator::allocate_partition(partition_constraints p_constraints) {
     return std::move(replicas).finish();
 }
 
+/**
+ * Check cluster-wide limits on total partition count vs available
+ * system resources.  This is the 'sanity' check that the user doesn't
+ * try and create a million partitions with only 8192 file handles, etc.
+ *
+ * We intentionally do this coarse global check rather than trying to
+ * carefully fit partitions into any available disk space etc, to avoid getting
+ * into situations where the partitions *only just* fit.  Partitions
+ * need to fit into the worst case node size, worst case node memory,
+ * to enable migrating partitions around during failures/rebalance/
+ * decommissions.
+ */
+std::error_code partition_allocator::check_cluster_limits(
+  allocation_request const& request) const {
+    if (_members.local().all_brokers().empty()) {
+        // Empty members table, we're probably running in a unit test
+        return errc::success;
+    }
+    // Calculate how many partition-replicas already exist, so that we can
+    // check if the new topic would take us past any limits.
+    uint64_t existent_partitions{0};
+    for (const auto& i : _state->allocation_nodes()) {
+        existent_partitions += uint64_t(i.second->allocated_partitions());
+    }
+
+    // Partition-replicas requested
+    uint64_t create_count{0};
+    for (const auto& i : request.partitions) {
+        create_count += uint64_t(i.replication_factor);
+    }
+
+    uint64_t proposed_total_partitions = existent_partitions + create_count;
+
+    // Gather information about system-wide resource sizes
+    uint32_t min_core_count = 0;
+    uint64_t min_memory_bytes = 0;
+    uint64_t min_disk_bytes = 0;
+    auto all_brokers = _members.local().all_brokers();
+    for (const auto& b : all_brokers) {
+        if (min_core_count == 0) {
+            min_core_count = b->properties().cores;
+        } else {
+            min_core_count = std::min(min_core_count, b->properties().cores);
+        }
+
+        // In redpanda <= 21.11.x, available_memory_gb and available_disk_gb
+        // are not populated.  If they're zero we skip the check later.
+        if (min_memory_bytes == 0) {
+            min_memory_bytes = b->properties().available_memory_gb * 1_GiB;
+        } else if (b->properties().available_memory_gb > 0) {
+            min_memory_bytes = std::min(
+              min_memory_bytes, b->properties().available_memory_gb * 1_GiB);
+        }
+
+        if (min_disk_bytes == 0) {
+            min_disk_bytes = b->properties().available_disk_gb * 1_GiB;
+        } else if (b->properties().available_disk_gb > 0) {
+            min_disk_bytes = std::min(
+              min_disk_bytes, b->properties().available_disk_gb * 1_GiB);
+        }
+    }
+
+    // The effective values are the node count times the smallest node's
+    // resources: this avoids wrongly assuming the system will handle partition
+    // counts that only fit when scheduled onto certain nodes.
```

Comment:
I am not sure if that is the valid assumption. This basically makes using heterogeneous clusters pointless. I think we should not make global decisions but rather use per node constraints to allocate partitions. Approach that is used here assumes that replication factor is equal to the number of nodes. And what if that is not the case, f.e. imagine a 6 nodes cluster with 3 small and 3 large nodes. This approach limits the cluster capacity to the capacity of 6 small nodes.

Reply:
This kind of situation highlights what's tricky about heterogenous clusters in general: if we have three large and three small nodes, we have to be super careful about ever decommissioning a node, because if it's one of the large ones then when it sheds partitions to the small node, it'll potentially bring the whole cluster down by overloading the small node.

We could/should improve decom process and partition movement process to make that kind of thing safe, but today those safety guards don't exist, so the safe (if slightly unsatisfying) move is to treat the cluster like it's all made of small nodes, while leaving the ability for our field team to modify these tunables to waive limits if necessary.

Reply:
Actually we have safety guarantees in place, all of the automatic partitions movements are made based on the partition allocator decisions. i.e. decommissioning/rebalancing will never succeed if there is not enough resources on the target nodes. 

We currently only decided based on the core counts but adding disk/fds/memory constraints would make allocator to decide based on complete set of information

Reply:
Interesting thread. Local decisions are more simple, but usually, longer term, it takes coordination to optimize space usage and allow for desired fault tolerance (e.g. guarantee a node can be lost and replication can still be rebuilt). Heterogeneous clusters are important, but usually are a problem that get solved later on (it often takes 5-10 years to iron out all of the higher priority issues in a distributed consensus storage system like this)